src/jit/decomposelongs.cpp

```diff
@@ -156,6 +156,40 @@ void DecomposeLongs::DecomposeNode(GenTree** ppTree, Compiler::fgWalkData* data)
 
     if (tree->TypeGet() != TYP_LONG)
     {
+        // For long -> int/short/byte conversions, if overflow checking is not needed,
+        // the upper 32 bits can be simply dropped. If it is not dropped here, codegen will 
+        // generate unnecessary instructions to compute the upper 32 bits.
+        //assert(!tree->gtOverflow());
+        if (tree->OperGet() == GT_CAST && !tree->gtOverflow())
+        {
+            var_types srcType = tree->AsCast()->CastFromType();
+            var_types dstType = tree->AsCast()->CastToType();
+
+            if (varTypeIsLong(srcType) && varTypeIsIntegral(dstType))
+            {
+                JITDUMP("[Decompose Long to int]:");
+                JITDUMP("Before:\n");
+                DISPTREE(tree);
+
+                assert(genTypeSize(srcType) > genTypeSize(dstType));
+                assert(tree->gtOp.gtOp1->OperGet() == GT_LONG);
+
+                GenTree* gtLong = tree->gtOp.gtOp1;
+                GenTree* highOpd = gtLong->gtOp.gtOp2;
+                GenTree* lowOpd = gtLong->gtOp.gtOp1;
+                GenTreeStmt* curStmt = m_compiler->compCurStmt->AsStmt();
+                unsigned dstSize = genTypeSize(dstType);
+                unsigned srcSize = genTypeSize(srcType);
+
+                m_compiler->fgSnipNode(curStmt, gtLong);
+                m_compiler->fgDeleteTreeFromList(curStmt, highOpd);
```

Comment:
>  Can it be simply done by checking the side effect flag on 'highOpd` node?

Yes, side-effect flags apply to the entire tree, not just the node. 

Assuming that the side-effect flags are correct in the first place, decomposition doesn't always get this right currently...